core/src/trezor/loop.py

```diff
@@ -421,3 +443,118 @@ def _schedule_take(self, taker: Task) -> None:
+    >>> except loop.TaskClosed:
+    >>>     print("task was closed before it could finish")
+    >>>
+    >>> # meanwhile, on the other side of town...
+    >>> controller.close()
+
+    Task is spawned only once. Multiple attempts to `await spawned_object` will return
+    the original return value (or raise the original exception).
+    """
+
+    def __init__(self, task: Task) -> None:
+        self.task = task
+        self.callback = None  # type: Optional[Task]
+        self.finalizer_callback = None  # type: Optional[Callable[["spawn"], None]]
+        self.finished = False
+        self.return_value = None  # type: Any
+
+        # schedule task immediately
+        if __debug__:
+            log.debug(__name__, "spawn new task: %s", task)
+        schedule(task, finalizer=self._finalize)
+
+    def _finalize(self, task: Task, value: Any) -> None:
+        # sanity check: make sure finalizer is for our task
+        assert task is self.task
+        # sanity check: make sure finalizer is not called more than once
+        assert self.finished is False
+
+        # now we are truly finished
+        self.finished = True
+        if isinstance(value, GeneratorExit):
+            # coerce GeneratorExit to a catchable TaskClosed
+            self.return_value = TASK_CLOSED
+        else:
+            self.return_value = value
+
+        if self.callback is not None:
+            schedule(self.callback, self.return_value)
+            self.callback = None
+        if self.finalizer_callback is not None:
+            self.finalizer_callback(self)
+
+    def __iter__(self) -> Task:  # type: ignore
+        if self.finished:
+            # exit immediately if we already have a return value
+            if isinstance(self.return_value, BaseException):
+                raise self.return_value
+            else:
+                return self.return_value
+
+        try:
+            return (yield self)
+        except BaseException:
+            # Clear out the callback. Otherwise we would raise the exception into it,
+            # AND schedule it with the closing value of the child task.
+            self.callback = None
+            assert self.task is not this_task  # closing parent from child :(
+            close(self.task)
+            raise
+
+    def handle(self, caller: Task) -> None:
+        # the same spawn should not be awaited multiple times
+        assert self.callback is None
+        self.callback = caller
+
+    def close(self) -> None:
+        """Shut down the spawned task.
+
+        If another caller is awaiting its result it will get a TaskClosed exception.
+        If the task was already finished, the call has no effect.
+        """
+        if not self.finished:
+            if __debug__:
+                log.debug(__name__, "close spawned task: %s", self.task)
+            close(self.task)
+
+    def on_close(self, finalizer_callback: Callable[["spawn"], None]) -> None:
```

Comment:
The function name here is confusing. This function is not called upon close, it is used to set the callback which should be called upon close. How about `set_finalizer()`?

Reply:
done (in a series of fixups)